ext/pdo/pdo_stmt.c

```diff
@@ -1342,12 +1286,15 @@ PHP_METHOD(PDOStatement, fetchAll)
 		}
 	}
 
-	do_fetch_opt_finish(stmt, 0);
-
 	/* Restore defaults which were changed by PDO_FETCH_CLASS mode */
-	stmt->fetch.cls.ce = old_ce;
-	ZVAL_COPY_VALUE(&stmt->fetch.cls.ctor_args, &old_ctor_args);
-	stmt->fetch.cls.fci.param_count = old_arg_count;
+	if (!is_default_fetch_mode && fetch_mode == PDO_FETCH_CLASS) {
+reset_old_ctor_args:
+		if (stmt->fetch.cls.ctor_args != NULL) {
+			zend_hash_release(stmt->fetch.cls.ctor_args);
```

Comment:
Same question: under what circumstances can this actually be freed? Intuitively I don't see it. The refcounting looks not useful? That should also simplify all the reasoning about the error paths etc

Reply:
Yeah turns out not need, this simplifies a lot.